NEW Add group syncing with automated DN mapping configuration.

mateusz opened the pull request:
Comes with an "LDAPAllSyncJob" for a full group+member sync.

mateusz committed:
NEW Add group syncing with automated DN mapping configuration.

Comes with an "LDAPAllSyncJob" for a full group+member sync.

halkyon commented on code/extensions/LDAPGroupExtension.php:
```diff
@@ -30,6 +41,19 @@ class LDAPGroupExtension extends DataExtension {
+		if ($this->owner->IsImportedFromLDAP) {
+			$fields->addFieldToTab('Root.Members', new LiteralField(
+				'Caution',
+				'<p class="message warning">Caution: LDAP group mapping is maintained automatically on this group. ' .
```
`_t()`?

mateusz replied:
Added all missing `_t`s

halkyon commented on code/extensions/LDAPGroupExtension.php:
```diff
@@ -30,6 +41,19 @@ class LDAPGroupExtension extends DataExtension {
 	public function updateCMSFields(FieldList $fields) {
+		$fields->addFieldToTab('Root.Members', new ReadonlyField('GUID'));
+		$fields->addFieldToTab('Root.Members', new ReadonlyField('DN'));
+		$fields->addFieldToTab('Root.Members', new ReadonlyField('IsImportedFromLDAP', 'Is group imported from LDAP/AD?'));
```
`_t()`? Probably check if the `LDAPMemberExtension` has this similar string localised as well.

halkyon commented:
Looks good, I added a few specific comments. Only other thing is we need documentation:
- How to configure the job, and regenerate time etc. There's already some docs, but add to them
- How to configure the group sync

Also, does it make sense to have a specific queued job for syncing groups only? Perhaps not, but I'm just thinking aloud here.

halkyon commented on code/tasks/LDAPGroupSyncTask.php:
```diff
@@ -0,0 +1,140 @@
+				));
+			}
+
+			$group->Title = $data['samaccountname'];
```
Is there an attribute that contains the human-readable name of the group? My thinking is use that attribute, if there's a value that isn't empty, and fallback to `samaccountname` if not.

mateusz replied:
So the `name` is now used as `Title` (this is the group name in AD). `samaccountname` was shifted into `Code` (this is Windows-2000 account name, akin to Member's Username). Also added `description` sync, since we have that fields on both sides too.

halkyon commented on code/tasks/LDAPGroupSyncTask.php:
```diff
@@ -0,0 +1,140 @@
+
+			// Delete stale mappings.
+			$hasCorrectMapping = false;
+			foreach ($group->LDAPGroupMappings() as $mapping) {
```
I'm a little confused what you're doing here. What is `hasCorrectMapping`? Could you explain it a bit in the comment?

mateusz replied:
Added more comments.

mateusz committed:
BUG Date was not set on LDAP user sync.

mateusz committed:
Mark LDAP-synced fields or make read-only. Move metadata to LDAP tab.

mateusz commented:
@halkyon I have changed the CMS UI to mark LDAP-synced fields in Member, and make them read-only in Groups, and also moved the metadata out into an LDAP tab because it was getting messy. There is a warning now too so people immediately see what is synced from LDAP. Makes sense?

mateusz commented:
Good point on docs - updated.

Regarding the sync jobs, I figured it doesn't make sense to have two separate jobs because to truly sync all groups we also need to sync users to get the memberships right. They also need to be run in order, and achieving this with two separate jobs running from queuedjobs could be tricky if the jobs take different amounts of time and start drifting. Hence one job to rule them all. I can add an extra `LDAPGroupSyncJob` if you think it's needed?

halkyon commented:
Makes sense. I'm merging it :)
